static/app/stores/persistedStore.tsx

```diff
@@ -0,0 +1,121 @@
+
+import OrganizationStore from './organizationStore';
+import {useLegacyStore} from './useLegacyStore';
+
+type PersistedStore = Readonly<{
+  onboarding: Partial<OnboardingState> | null;
+}>;
+
+const DefaultPersistedStore: PersistedStore = {
+  onboarding: null,
+};
+
+type PersistedStoreContextValue = [
+  PersistedStore | null,
+  React.Dispatch<React.SetStateAction<PersistedStore | null>>
+];
+
+export const PersistedStoreContext = createContext<PersistedStoreContextValue | null>(
+  null
+);
+
+function usePersistedStore(): PersistedStoreContextValue {
+  const context = useContext(PersistedStoreContext);
+
+  if (!context) {
+    throw new Error('usePersistedStore was called outside of PersistedStoreProvider');
+  }
+
+  return context;
+}
+
+// Client-only state with TTL persisted on the server side in a redis store.
+export function PersistedStoreProvider(props: {children: React.ReactNode}) {
+  const [state, setState] = useState<PersistedStore | null>(null);
+
+  const api = useApi();
+  const {organization} = useLegacyStore(OrganizationStore);
+
+  useEffect(() => {
+    if (!organization) {
+      return undefined;
+    }
+
+    let shouldCancelRequest = false;
+
+    api
+      .requestPromise(`/organizations/${organization.slug}/client-state/`)
+      .then((response: PersistedStore) => {
+        if (shouldCancelRequest) {
+          return;
+        }
+
+        setState(oldState => ({...oldState, ...response}));
+      });
+
+    return () => {
+      shouldCancelRequest = true;
+    };
+  }, [organization]);
+
+  return (
+    <PersistedStoreContext.Provider value={[state, setState]}>
+      {props.children}
+    </PersistedStoreContext.Provider>
+  );
+}
+
+type UsePersistedCategory<T> = [T | null, (nextState: T | null) => void];
+export function usePersistedStoreCategory<C extends keyof PersistedStore>(
+  category: C
+): UsePersistedCategory<PersistedStore[C]> {
+  type T = PersistedStore[C];
+  const api = useApi();
+  const organization = useOrganization();
+  const [state, setState] = usePersistedStore();
+
+  const setCategoryState = useCallback(
+    (val: T | null) => {
+      setState(oldState => ({...(oldState || DefaultPersistedStore), [category]: val}));
+
+      // If a state is set with null, we can clear it from the server.
+      if (val === null) {
+        api.requestPromise(
+          `/organizations/${organization.slug}/client-state/${category}/`,
+          {
+            method: 'DELETE',
+          }
+        );
+        return;
+      }
+
+      // Else we want to sync our state with the server
+      api.requestPromise(
+        `/organizations/${organization.slug}/client-state/${category}/`,
+        {
+          method: 'PUT',
+          data: val,
+        }
+      );
+    },
+    [category, organization]
+  );
+
+  const stableState: UsePersistedCategory<T> = useMemo(() => {
+    return [state?.[category] ?? null, setCategoryState];
+  }, [state?.[category], setCategoryState]);
```

Comment:
If you decide to update  based on my [suggestion](https://github.com/getsentry/sentry/pull/33489/files/3302b98364865cba52eaf54096354f5f9867a3c4#r848350146), you can drop safe property access and just keep nullish coalescing
```suggestion
  const stableState: UsePersistedCategory<T> = useMemo(() => {
    return [state.[category] ?? null, setCategoryState];
  }, [state.[category], setCategoryState]);
```